test/parallel/test-runner-misc.js

```diff
@@ -0,0 +1,33 @@
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const { spawnSync } = require('child_process');
+const { setTimeout } = require('timers/promises');
+
+if (process.argv[2] === 'child') {
```

Comment:
Why use a subprocess here? Wouldn't it be simpler to test in the same process?

```js
test(async () => {
  let testSignal;
  await assert.rejects(test({ timeout: 10 }, ({ signal }) => {
    testSignal = signal;
    assert.strictEqual(signal.aborted, false);
    return new Promise(() => {});
  }), { code: 'ERR_TEST_FAILURE' });

  setTimeout(50, common.mustCall(() => assert.strictEqual(testSignal?.aborted, true)));
});
```

Otherwise we might as well do a message test instead.

Reply:
`test()` never rejects for the same reason discussed above - the tap output will not be reliable otherwise

Reply:
let me rephrase: it is not just about the TAP output, but about what is the essential API of a test runner and what a test runner is expected to do,
by definition, a test catches errors/rejections and does not throw when it caught one - it reports a failure in a standard way so whoever is looking (ci/human) can know the test failed. 
if you simply throw any error you catch - there is no real reason to use a test runner (correct me if I am missing anything?)

the only difference here is that this error is not generated within the test/by the test author - but by the test runner.

the question to be asked is "is a timeout considered a test failure?" if yes then it should behave the same as `test(Promise.reject(new Error())` which definitly should not reject

>  when running on the CI, users will use --test CLI flag

I disagree about only relating to that as the test runner. users should be able to also run a single test file and get valid output, or run multiple files using any logic wanted.

Reply:
Would rejecting the promise be at odd with any of those goals really? I don't see a reason why we can't have both: `test(…)` returns a rejected promise, and we output valid TAP output on stdout.

Reply:
I tested this piece of code with a few frameworks (syntax adjusted a little per framework):

```js
test('1', async() => {
    await test('1.1', Promise.reject(new Error()));
    await test('1.2', Promise.reject(new Error()));
})
```
What you suggest (unless I misunderstood) is that only tests 1, 1.1 need to run - and I think users expect all three 1, 1.1, 1.2 to run

- playwright - ran all three tests (`test` does not return a promise)
- mocha - ran all three tests (`test` does not return a promise)
- jest - this isn't even possible - you must define all tests synchronously (`Returning a Promise from "describe" is not supported. Tests must be defined synchronously.`) in addition `test` does not return a promise
- tape - ran all three tests (`test` does not return a promise)
- tap - ran all three tests (this is the only framework where `test` actually returns a promise)

Reply:
I can see why test runners that don't return a promise would do that, it's weird to me that a promise resolving when the previous test fails. Anyway, if that's the consensus, let's roll with it, although it makes testing the test runner a bit trickier.

Reply:
Yeah, the fact that a test framework returns a promise can be confusing, see this issue as well:
https://github.com/nodejs/help/issues/3837

Reply:
this makes me think that a test runner that performs IOC should probably not return a promise